{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cff6a8c4-552b-4ba5-8a30-fb71f0423188",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Renamed GSM6668947_P18_T.enes.tsv.gz to GSM6668947_P18_T.genes.tsv.gz\n",
    "# Renamed GSM6668932_P5_TN.features.tsv.gz to GSM6668932_P5_TN.genes.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4496f196-ee52-428e-b405-ee32d2de2d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "from copy import deepcopy\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d13ba1-4741-4ff3-8934-009621767ee5",
   "metadata": {},
   "source": [
    "# Files and prefixes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1497ff5-eab5-4a1a-9f7d-076e775c4d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = \"../../original_datasets/ICI/GSE216329_RAW\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2e6b997e-7da5-4aa9-99d6-3673bb9cc3ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "144"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = os.listdir(base_path)\n",
    "files = [x for x in files if '.gz' in x]\n",
    "len(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "670accc2-e1da-4d6b-b780-1481b6950ea8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "48"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prefixes = np.unique([x.split('.')[0] for x in files]).tolist()\n",
    "len(prefixes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f198d193-01ec-4c88-8185-1633684822bc",
   "metadata": {},
   "source": [
    "# Construct adata per file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "876e9dac-991b-4ff1-a79b-5a4e44e113d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████| 48/48 [00:27<00:00,  1.74it/s]\n"
     ]
    }
   ],
   "source": [
    "# Build a mapping from gene symbol -> Ensembl ID using the files that contain both.\n",
    "# (Needed because some samples provide only symbols, while others provide only Ensembl IDs.)\n",
    "_sym_to_ens = {}\n",
    "_ens_to_sym = {}\n",
    "\n",
    "for prefix in prefixes:\n",
    "    gene_file = f'{base_path}/{prefix}.genes.tsv.gz'\n",
    "    gr = pd.read_csv(gene_file, sep='\\t', header=None)\n",
    "    if gr.shape[1] == 3:\n",
    "        gene_id = gr.iloc[:, 0].astype(str)\n",
    "        gene_sym = gr.iloc[:, 1].astype(str)\n",
    "        for e, s in zip(gene_id, gene_sym):\n",
    "            if e.startswith('ENSG') and (not s.startswith('ENSG')):\n",
    "                _sym_to_ens.setdefault(s, set()).add(e)\n",
    "                _ens_to_sym.setdefault(e, set()).add(s)\n",
    "\n",
    "sym2ens = {s: next(iter(v)) for s, v in _sym_to_ens.items() if len(v) == 1}\n",
    "ens2sym = {e: next(iter(v)) for e, v in _ens_to_sym.items() if len(v) == 1}\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in tqdm(prefixes):\n",
    "\n",
    "    # File names\n",
    "    gene_file = f'{base_path}/{prefix}.genes.tsv.gz'\n",
    "    barcode_file = f'{base_path}/{prefix}.barcodes.tsv.gz'\n",
    "    mtx_file = f'{base_path}/{prefix}.matrix.mtx.gz'\n",
    "\n",
    "    # Genes\n",
    "    genes_raw = pd.read_csv(gene_file, sep='\\t', header=None)\n",
    "\n",
    "    if genes_raw.shape[1] == 3:\n",
    "        genes = genes_raw.copy()\n",
    "        genes.columns = ['gene_name', 'gene_symbol', 'data']\n",
    "\n",
    "    elif genes_raw.shape[1] == 2:\n",
    "        c0 = genes_raw.iloc[:, 0].astype(str)\n",
    "        c1 = genes_raw.iloc[:, 1].astype(str)\n",
    "\n",
    "        # Case A: Ensembl present (sometimes mixed with non-gene features in the same file)\n",
    "        # We keep only Ensembl IDs later via `valid_gene`.\n",
    "        if (c0 == c1).all() and c0.str.startswith('ENSG').any():\n",
    "            gene_name = c0\n",
    "            gene_symbol = c0.map(lambda e: ens2sym.get(e, e))\n",
    "\n",
    "        # Case B: Symbol-only (both columns are symbols)\n",
    "        elif (not c0.str.startswith('ENSG').any()) and (c0 == c1).all():\n",
    "            gene_symbol = c0\n",
    "            # normalize RP11. -> RP11- only if present\n",
    "            if gene_symbol.str.contains(r'^RP\\d+\\.', regex=True).any():\n",
    "                gene_symbol = gene_symbol.str.replace(r'^(RP\\d+)\\.', r'\\1-', regex=True)\n",
    "            gene_name = gene_symbol.map(lambda s: sym2ens.get(s, pd.NA))\n",
    "\n",
    "        # Case C: Ensembl + symbol (2 columns)\n",
    "        elif c0.str.startswith('ENSG').all() and (not c1.str.startswith('ENSG').all()):\n",
    "            gene_name = c0\n",
    "            gene_symbol = c1\n",
    "\n",
    "        else:\n",
    "            raise ValueError(f\"Unexpected 2-col genes format for {prefix}: first rows={genes_raw.head(3).values.tolist()}\")\n",
    "\n",
    "        genes = pd.DataFrame({'gene_name': gene_name, 'gene_symbol': gene_symbol, 'data': pd.NA})\n",
    "\n",
    "    else:\n",
    "        raise ValueError(f\"Unexpected gene file format: {gene_file} shape={genes_raw.shape}\")\n",
    "\n",
    "    genes['gene_name'] = genes['gene_name'].astype(str)\n",
    "    genes['gene_symbol'] = genes['gene_symbol'].astype(str)\n",
    "\n",
    "    # Drop genes that we could not map to Ensembl (only relevant for symbol-only files)\n",
    "    valid_gene = genes['gene_name'].notna() & (genes['gene_name'].str.lower() != 'nan') & (genes['gene_name'] != '') & (genes['gene_name'].str.startswith('ENSG'))\n",
    "\n",
    "    # Barcodes\n",
    "    barcodes = pd.read_csv(\n",
    "        barcode_file,\n",
    "        sep='\\t',\n",
    "        header=None,\n",
    "        names=['sample_id']\n",
    "    )\n",
    "\n",
    "    # Matrix\n",
    "    mtx = sc.read_mtx(mtx_file).T\n",
    "\n",
    "    # If feature_type is present, keep only Gene Expression rows.\n",
    "    if genes['data'].notna().any() and (genes['data'] == 'Gene Expression').any():\n",
    "        keep = (genes['data'] == 'Gene Expression').values\n",
    "        genes = genes.loc[keep].reset_index(drop=True)\n",
    "        mtx = mtx[:, keep].copy()\n",
    "        valid_gene = valid_gene.loc[keep].reset_index(drop=True)\n",
    "\n",
    "    # Apply mapping filter (and keep matrix aligned)\n",
    "    if not bool(valid_gene.all()):\n",
    "        genes = genes.loc[valid_gene].reset_index(drop=True)\n",
    "        mtx = mtx[:, valid_gene.values].copy()\n",
    "\n",
    "    # Assertions\n",
    "    assert barcodes.shape[1] == 1\n",
    "    assert (len(barcodes), len(genes)) == mtx.shape\n",
    "    assert len(genes['gene_name'].unique()) == len(genes)\n",
    "\n",
    "    # Set data: align across samples by Ensembl ID\n",
    "    mtx.var = deepcopy(genes)\n",
    "    mtx.var_names = mtx.var['gene_name'].copy()\n",
    "\n",
    "    # Make obs_names unique across samples\n",
    "    mtx.obs_names = [f\"{x}-{prefix}\" for x in barcodes['sample_id'].astype(str).tolist()]\n",
    "    mtx.obs['sample_id'] = barcodes['sample_id'].astype(str).tolist()\n",
    "\n",
    "    # Metadata\n",
    "    geo, patient_id, t = prefix.split('_')\n",
    "    mtx.obs['prefix'] = prefix\n",
    "    mtx.obs['GEO'] = geo\n",
    "    mtx.obs['patient_id'] = patient_id\n",
    "    mtx.obs['T'] = t\n",
    "\n",
    "    adatas.append(mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4882d7af-5c54-4461-92f3-b0af3ddc2567",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersection_genes = set(adatas[0].var_names)\n",
    "for a in adatas[1:]:\n",
    "    intersection_genes &= set(a.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "40486d1a-e4f6-477d-89fc-04be1bf4d589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20341"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_genes = list(intersection_genes)\n",
    "len(common_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9eb1cd79-8ef7-4978-80a8-1b3c69fd36bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example (after the fix above, var_names are gene symbols):\n",
    "# i = 0\n",
    "# adatas[i][:, ['RP11-34P13.3', 'FAM138A']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "31f48cdc-5810-4cd7-a73a-3983024a3d15",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, _adata in enumerate(adatas):\n",
    "    adatas[i] = adatas[i][:, common_genes]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edbbbb69-3461-45a0-9f4e-dba604609c23",
   "metadata": {},
   "source": [
    "# Assertions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "115507b8-48da-4698-b064-ef2cccf400db",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = adatas[0].var_names.tolist()\n",
    "for i, _adata in enumerate(adatas):\n",
    "    assert len(_adata.var_names.tolist()) == len(gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "82580e1c-3b83-4a36-925e-835da6b2a5c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20341"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(_adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8f63052b-3050-4dc0-8acf-1a9b14fd89a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20341"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f0a8bf8e-2c5a-46cc-9898-35fc38b10b69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000111671', 'ENSG00000183260', 'ENSG00000173826',\n",
       "       'ENSG00000169393', 'ENSG00000253159', 'ENSG00000229563',\n",
       "       'ENSG00000136297', 'ENSG00000052344', 'ENSG00000115474',\n",
       "       'ENSG00000138092',\n",
       "       ...\n",
       "       'ENSG00000215374', 'ENSG00000166126', 'ENSG00000141579',\n",
       "       'ENSG00000136936', 'ENSG00000224511', 'ENSG00000112893',\n",
       "       'ENSG00000117305', 'ENSG00000239590', 'ENSG00000166016',\n",
       "       'ENSG00000179796'],\n",
       "      dtype='object', name='gene_name', length=20341)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_adata.var_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f01e0ac-629b-40e2-b557-7a1b37795790",
   "metadata": {},
   "source": [
    "# Concatenate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8f906e41-343b-4250-810c-4a93fcd5bfad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "170295c1-c53c-48c3-a625-6b36d5e40870",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/SE/miniconda3/envs/trials/lib/python3.10/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = ad.concat(adatas)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17412247-f616-4de1-bacb-0864556f3c39",
   "metadata": {},
   "source": [
    "# Add Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ca382a76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sample_id',\n",
       " 'prefix',\n",
       " 'GEO',\n",
       " 'patient_id',\n",
       " 'T',\n",
       " 'title',\n",
       " 'source',\n",
       " 'organism',\n",
       " 'ch1__cell type',\n",
       " 'ch1__tissue',\n",
       " 'ch1__disease',\n",
       " 'ch1__condition',\n",
       " 'library_strategy',\n",
       " 'library_source',\n",
       " 'library_selection']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "import tarfile\n",
    "import urllib.request\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "gse = 'GSE216329'\n",
    "url = f'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE216nnn/{gse}/miniml/{gse}_family.xml.tgz'\n",
    "\n",
    "with urllib.request.urlopen(url) as r:\n",
    "    data = r.read()\n",
    "\n",
    "with tarfile.open(fileobj=io.BytesIO(data), mode='r:gz') as tf:\n",
    "    xml_member = [m for m in tf.getmembers() if m.name.endswith('.xml')][0]\n",
    "    xml_bytes = tf.extractfile(xml_member).read()\n",
    "\n",
    "root = ET.fromstring(xml_bytes)\n",
    "\n",
    "strip_ns = lambda tag: tag.split('}', 1)[-1] if '}' in tag else tag\n",
    "\n",
    "rows = []\n",
    "for sample in root.iter():\n",
    "    if strip_ns(sample.tag) != 'Sample':\n",
    "        continue\n",
    "\n",
    "    rec = {}\n",
    "    for child in sample:\n",
    "        if strip_ns(child.tag) == 'Accession':\n",
    "            rec['GEO'] = (child.text or '').strip()\n",
    "            break\n",
    "    if not rec.get('GEO', '').startswith('GSM'):\n",
    "        continue\n",
    "\n",
    "    for child in sample.iter():\n",
    "        t = strip_ns(child.tag)\n",
    "        if t in {'Title', 'Source', 'Organism', 'Library-Strategy', 'Library-Source', 'Library-Selection'} and child.text:\n",
    "            rec[t.lower().replace('-', '_')] = child.text.strip()\n",
    "        if t == 'Characteristics':\n",
    "            k = child.attrib.get('tag', None)\n",
    "            v = (child.text or '').strip()\n",
    "            if k:\n",
    "                rec[f'ch1__{k}'] = v\n",
    "\n",
    "    rows.append(rec)\n",
    "\n",
    "sample_meta = pd.DataFrame(rows).drop_duplicates(subset=['GEO']).set_index('GEO')\n",
    "adata.obs = adata.obs.join(sample_meta, on='GEO')\n",
    "\n",
    "adata.obs.columns.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bad854-4205-4c14-9a95-34ee2493ac3d",
   "metadata": {},
   "source": [
    "# Rename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7cf428c6-384d-4e45-b0a7-57cfc7c5fb34",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs_names_make_unique()\n",
    "adata.obs.rename(\n",
    "    columns={\n",
    "        'ch1__cell_type': 'cell_type',\n",
    "        'ch1__disease': 'disease',\n",
    "        'ch1__condition':'condition',\n",
    "        'ch1__tissue': 'tissue'\n",
    "    }, \n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "47fa8b2d-60e8-4612-b137-ca5adf778bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 222678 × 20341\n",
       "    obs: 'sample_id', 'prefix', 'GEO', 'patient_id', 'T', 'title', 'source', 'organism', 'ch1__cell type', 'tissue', 'disease', 'condition', 'library_strategy', 'library_source', 'library_selection'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "941621ff-6636-48ef-850d-4b3a2f9b62fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.crosstab(adata.obs['disease'], adata.obs['patient_id']).to_csv('patient_disease_crosstab.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "100b4070-a84b-4b46-aa95-e4b7cc2ec109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGCTTATCG-1-GSM6668925_P1_T      On treatment\n",
       "AAACCTGAGGATGCGT-1-GSM6668925_P1_T      On treatment\n",
       "AAACCTGAGGTGATAT-1-GSM6668925_P1_T      On treatment\n",
       "AAACCTGCACATAACC-1-GSM6668925_P1_T      On treatment\n",
       "AAACCTGCACCTCGTT-1-GSM6668925_P1_T      On treatment\n",
       "                                            ...     \n",
       "TTTATGCCATGCCCGA.1-GSM6668972_P24_TN        Baseline\n",
       "TTTATGCGTCTTCTCG.1-GSM6668972_P24_TN        Baseline\n",
       "TTTCCTCTCGGATGTT.1-GSM6668972_P24_TN        Baseline\n",
       "TTTGTCAAGGAGTTTA.1-GSM6668972_P24_TN        Baseline\n",
       "TTTGTCAGTGGTCTCG.1-GSM6668972_P24_TN        Baseline\n",
       "Name: condition, Length: 222678, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['condition']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0c32d1f3-9b7f-479c-a652-960d853ea2b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGCTTATCG-1-GSM6668925_P1_T      Arthritis\n",
       "AAACCTGAGGATGCGT-1-GSM6668925_P1_T      Arthritis\n",
       "AAACCTGAGGTGATAT-1-GSM6668925_P1_T      Arthritis\n",
       "AAACCTGCACATAACC-1-GSM6668925_P1_T      Arthritis\n",
       "AAACCTGCACCTCGTT-1-GSM6668925_P1_T      Arthritis\n",
       "                                          ...    \n",
       "TTTATGCCATGCCCGA.1-GSM6668972_P24_TN      No irAE\n",
       "TTTATGCGTCTTCTCG.1-GSM6668972_P24_TN      No irAE\n",
       "TTTCCTCTCGGATGTT.1-GSM6668972_P24_TN      No irAE\n",
       "TTTGTCAAGGAGTTTA.1-GSM6668972_P24_TN      No irAE\n",
       "TTTGTCAGTGGTCTCG.1-GSM6668972_P24_TN      No irAE\n",
       "Name: disease, Length: 222678, dtype: object"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['disease']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7277dfa9-12e7-46f3-9dc5-4c056485a0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ens = pd.Series(adata.var_names.astype(str), index=adata.var.index)\n",
    "ens_nover = ens.str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
    "\n",
    "sym = ens_nover.map(lambda e: ens2sym.get(e, pd.NA))\n",
    "\n",
    "adata.var[\"gene_symbol\"] = sym\n",
    "\n",
    "adata.var[\"gene_id_ensembl\"] = ens_nover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5c05f268-1e23-40ab-92ea-3ab63a088410",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_id_ensembl</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000111671</th>\n",
       "      <td>SPSB2</td>\n",
       "      <td>ENSG00000111671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000183260</th>\n",
       "      <td>ABHD16B</td>\n",
       "      <td>ENSG00000183260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000173826</th>\n",
       "      <td>KCNH6</td>\n",
       "      <td>ENSG00000173826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000169393</th>\n",
       "      <td>ELSPBP1</td>\n",
       "      <td>ENSG00000169393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000253159</th>\n",
       "      <td>PCDHGA12</td>\n",
       "      <td>ENSG00000253159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000112893</th>\n",
       "      <td>MAN2A1</td>\n",
       "      <td>ENSG00000112893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000117305</th>\n",
       "      <td>HMGCL</td>\n",
       "      <td>ENSG00000117305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239590</th>\n",
       "      <td>OR1J4</td>\n",
       "      <td>ENSG00000239590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000166016</th>\n",
       "      <td>ABTB2</td>\n",
       "      <td>ENSG00000166016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000179796</th>\n",
       "      <td>LRRC3B</td>\n",
       "      <td>ENSG00000179796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20341 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_symbol  gene_id_ensembl\n",
       "gene_name                                   \n",
       "ENSG00000111671       SPSB2  ENSG00000111671\n",
       "ENSG00000183260     ABHD16B  ENSG00000183260\n",
       "ENSG00000173826       KCNH6  ENSG00000173826\n",
       "ENSG00000169393     ELSPBP1  ENSG00000169393\n",
       "ENSG00000253159    PCDHGA12  ENSG00000253159\n",
       "...                     ...              ...\n",
       "ENSG00000112893      MAN2A1  ENSG00000112893\n",
       "ENSG00000117305       HMGCL  ENSG00000117305\n",
       "ENSG00000239590       OR1J4  ENSG00000239590\n",
       "ENSG00000166016       ABTB2  ENSG00000166016\n",
       "ENSG00000179796      LRRC3B  ENSG00000179796\n",
       "\n",
       "[20341 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "df68ec21-8a2b-49c9-a1aa-558194611759",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16688.0, 0.0)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.max(), adata.X.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "568f2e50-7eae-4e45-aeef-1b440f7172f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('../../preprocessed_datasets/ICI_original_concat.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7ef73477-8b88-4c00-8c72-61ea20d7c940",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>disease</th>\n",
       "      <th>Arthritis</th>\n",
       "      <th>Colitis/ Neurotoxicity</th>\n",
       "      <th>No irAE</th>\n",
       "      <th>Pneumonitis</th>\n",
       "      <th>Thyroiditis</th>\n",
       "      <th>Thyroiditis/ Nephritis</th>\n",
       "      <th>Thyroiditis/ Neurotoxicity / Nephritis</th>\n",
       "      <th>Thyroiditis/ Neurotoxicity/ Nephritis</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P1</th>\n",
       "      <td>16346</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P2</th>\n",
       "      <td>8936</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3</th>\n",
       "      <td>11053</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21941</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22389</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>778</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P7</th>\n",
       "      <td>8320</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12150</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12842</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13894</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13517</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P12</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5638</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3069</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P14</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4625</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P15</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2299</td>\n",
       "      <td>711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P16</th>\n",
       "      <td>0</td>\n",
       "      <td>4684</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P17</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11880</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P18</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7937</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P19</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12228</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P20</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P21</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5143</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P22</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7037</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P23</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3088</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P24</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6022</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "disease     Arthritis  Colitis/ Neurotoxicity  No irAE  Pneumonitis  \\\n",
       "patient_id                                                            \n",
       "P1              16346                       0        0            0   \n",
       "P2               8936                       0        0            0   \n",
       "P3              11053                       0        0            0   \n",
       "P4                  0                       0        0            0   \n",
       "P5                  0                       0        0            0   \n",
       "P6                  0                       0        0            0   \n",
       "P7               8320                       0        0            0   \n",
       "P8                  0                       0        0        12150   \n",
       "P9                  0                       0    12842            0   \n",
       "P10                 0                       0        0        13894   \n",
       "P11                 0                       0        0        13517   \n",
       "P12                 0                       0        0         5638   \n",
       "P13                 0                       0        0         3069   \n",
       "P14                 0                       0        0         4625   \n",
       "P15                 0                       0        0            0   \n",
       "P16                 0                    4684        0            0   \n",
       "P17                 0                       0    11880            0   \n",
       "P18                 0                       0     7937            0   \n",
       "P19                 0                       0    12228            0   \n",
       "P20                 0                       0     6151            0   \n",
       "P21                 0                       0     5143            0   \n",
       "P22                 0                       0     7037            0   \n",
       "P23                 0                       0     3088            0   \n",
       "P24                 0                       0     6022            0   \n",
       "\n",
       "disease     Thyroiditis  Thyroiditis/ Nephritis  \\\n",
       "patient_id                                        \n",
       "P1                    0                       0   \n",
       "P2                    0                       0   \n",
       "P3                    0                       0   \n",
       "P4                21941                       0   \n",
       "P5                22389                       0   \n",
       "P6                    0                     778   \n",
       "P7                    0                       0   \n",
       "P8                    0                       0   \n",
       "P9                    0                       0   \n",
       "P10                   0                       0   \n",
       "P11                   0                       0   \n",
       "P12                   0                       0   \n",
       "P13                   0                       0   \n",
       "P14                   0                       0   \n",
       "P15                   0                       0   \n",
       "P16                   0                       0   \n",
       "P17                   0                       0   \n",
       "P18                   0                       0   \n",
       "P19                   0                       0   \n",
       "P20                   0                       0   \n",
       "P21                   0                       0   \n",
       "P22                   0                       0   \n",
       "P23                   0                       0   \n",
       "P24                   0                       0   \n",
       "\n",
       "disease     Thyroiditis/ Neurotoxicity / Nephritis  \\\n",
       "patient_id                                           \n",
       "P1                                               0   \n",
       "P2                                               0   \n",
       "P3                                               0   \n",
       "P4                                               0   \n",
       "P5                                               0   \n",
       "P6                                               0   \n",
       "P7                                               0   \n",
       "P8                                               0   \n",
       "P9                                               0   \n",
       "P10                                              0   \n",
       "P11                                              0   \n",
       "P12                                              0   \n",
       "P13                                              0   \n",
       "P14                                              0   \n",
       "P15                                           2299   \n",
       "P16                                              0   \n",
       "P17                                              0   \n",
       "P18                                              0   \n",
       "P19                                              0   \n",
       "P20                                              0   \n",
       "P21                                              0   \n",
       "P22                                              0   \n",
       "P23                                              0   \n",
       "P24                                              0   \n",
       "\n",
       "disease     Thyroiditis/ Neurotoxicity/ Nephritis  \n",
       "patient_id                                         \n",
       "P1                                              0  \n",
       "P2                                              0  \n",
       "P3                                              0  \n",
       "P4                                              0  \n",
       "P5                                              0  \n",
       "P6                                              0  \n",
       "P7                                              0  \n",
       "P8                                              0  \n",
       "P9                                              0  \n",
       "P10                                             0  \n",
       "P11                                             0  \n",
       "P12                                             0  \n",
       "P13                                             0  \n",
       "P14                                             0  \n",
       "P15                                           711  \n",
       "P16                                             0  \n",
       "P17                                             0  \n",
       "P18                                             0  \n",
       "P19                                             0  \n",
       "P20                                             0  \n",
       "P21                                             0  \n",
       "P22                                             0  \n",
       "P23                                             0  \n",
       "P24                                             0  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(adata.obs['disease'], adata.obs['patient_id']).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88fdf255-9647-4d83-99d7-eed1a09a1920",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (trials)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
