{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5b7cf2c-4444-48f1-aa6a-f0c02d3b0931",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import sys\n",
    "\n",
    "sys.path.append('../')\n",
    "import preprocessing_tools as pt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b2697086-4909-44a4-9727-cd8aa97be03a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"../../original_datasets/Tabula\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce0b09e1-8386-40d4-8ebe-143e2033aa3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sapiens = sc.read_h5ad(f'{data_path}/40f8b1a3-9f76-4ac4-8761-32078555ed4e.h5ad')\n",
    "adata_muris = sc.read_h5ad(f'{data_path}/b65b771c-baae-4b03-befc-3db63e6b9fd0.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f73468d9-dcdf-46ec-b5be-356d9b936af4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 24540 × 17943\n",
       "    obs: 'age', 'cell', 'free_annotation', 'method', 'donor_id', 'n_genes', 'subtissue', 'n_counts', 'louvain', 'leiden', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type'\n",
       "    uns: 'age_colors', 'citation', 'leiden', 'louvain', 'neighbors', 'organism', 'organism_ontology_term_id', 'pca', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_muris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "172d86ae-9376-4673-ab2a-c5dd53def06b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10.0, 8.976758)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_muris.X.max(), adata_sapiens.X.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8e6d6ef6-d176-436d-8f4a-c719abc5a06b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_muris.X = adata_muris.raw.X\n",
    "adata_sapiens.X = adata_sapiens.raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ffbfeead-d6d5-4f50-bce6-fb704ab2598f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<24540x17943 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 37375133 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_muris.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "555fd5c6-99ee-4a99-b881-3e2cca6079e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sapien_genes = adata_sapiens.var['feature_name'].tolist()\n",
    "muris_genes = adata_muris.var['feature_name'].tolist()\n",
    "\n",
    "sapiens_genes = [x.upper() for x in sapien_genes]\n",
    "muris_genes = [x.upper() for x in muris_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8d6d2f4e-f803-46ec-8475-43764271e17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap = [x for x in sapien_genes if x in muris_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fa0d9fbb-93dd-4ebb-810c-e0fb9e7a5163",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14671"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "02e12a22-7a0d-4c77-9cf2-546fe60982ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-12-12 02:43:46--  https://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt\n",
      "Resolving www.informatics.jax.org (www.informatics.jax.org)... 34.102.200.12\n",
      "Connecting to www.informatics.jax.org (www.informatics.jax.org)|34.102.200.12|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 15081658 (14M)\n",
      "Saving to: ‘HOM_MouseHumanSequence.rpt’\n",
      "\n",
      "HOM_MouseHumanSeque 100%[===================>]  14.38M  10.3MB/s    in 1.4s    \n",
      "\n",
      "2025-12-12 02:43:48 (10.3 MB/s) - ‘HOM_MouseHumanSequence.rpt’ saved [15081658/15081658]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget https://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a8e1a5d3-4f1c-424f-9a37-94bba61eb004",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "FILE_HOMOLOGY = Path(\"HOM_MouseHumanSequence.rpt\")\n",
    "mouse_tax, human_tax = \"10090\", \"9606\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "644bbe38-82e6-4758-b538-4cece5df4cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------------------------------------------\n",
    "# 1 · read MGI report ----------------------------------------------\n",
    "# ------------------------------------------------------------------\n",
    "cols = [\n",
    "    \"homology_id\", \"organism\", \"taxon_id\", \"symbol\", \"entrez_id\",\n",
    "    \"marker_type\", \"synonyms\", \"chromosome\", \"start\", \"end\", \"strand\",\n",
    "    \"homology_type\"\n",
    "]\n",
    "hom = pd.read_csv(FILE_HOMOLOGY, sep=\"\\t\", header=None,\n",
    "                  names=cols, usecols=range(len(cols)), dtype=str)\n",
    "\n",
    "# ------------------------------------------------------------------\n",
    "# 2 · keep groups that have BOTH species ---------------------------\n",
    "# ------------------------------------------------------------------\n",
    "has_mouse = hom[\"taxon_id\"] == mouse_tax\n",
    "has_human = hom[\"taxon_id\"] == human_tax\n",
    "good_ids  = hom.loc[has_mouse | has_human, \"homology_id\"].value_counts().index\n",
    "\n",
    "mh = hom[hom[\"homology_id\"].isin(good_ids)]\n",
    "\n",
    "# ------------------------------------------------------------------\n",
    "# 3 · pick ONE row per species in each group -----------------------\n",
    "# ------------------------------------------------------------------\n",
    "mouse_rows = (mh[mh[\"taxon_id\"] == mouse_tax]\n",
    "                .groupby(\"homology_id\").first())  # first = arbitrary but consistent\n",
    "human_rows = (mh[mh[\"taxon_id\"] == human_tax]\n",
    "                .groupby(\"homology_id\").first())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "29f534c0-a129-401c-9be9-94ca58269bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "orth = (mouse_rows[\"symbol\"]\n",
    "        .to_frame(\"mouse_symbol\")\n",
    "        .join(human_rows[\"symbol\"].to_frame(\"human_symbol\"),\n",
    "              how=\"inner\"))\n",
    "\n",
    "orth = orth.dropna().drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1ec6d165-3354-4d03-ae55-8a3a862397a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/SE/miniconda3/envs/trials/lib/python3.10/site-packages/anndata/utils.py:260: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD115-1', 'SNORD115-2', 'SNORD115-3', 'SNORD115-4', 'SNORD115-5']\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "for ad in (adata_muris, adata_sapiens):\n",
    "    ad.var_names = ad.var[\"feature_name\"].tolist()        # or whatever column holds symbols\n",
    "    ad.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d300d4f1-dabc-4d9b-a82a-3e393705c445",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) Filter mouse genes to those present in the ortholog table\n",
    "keep_mouse = orth[\"mouse_symbol\"]\n",
    "adata_muris = adata_muris[:, adata_muris.var_names.isin(keep_mouse)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aee7e4b4-66e5-4850-92a8-c781c813459c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2) Rename mouse genes to their human ortholog symbols\n",
    "mapper = dict(zip(orth[\"mouse_symbol\"], orth[\"human_symbol\"]))\n",
    "adata_muris.var_names = adata_muris.var_names.map(mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2805601e-be51-4925-a954-8cc001af4f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_muris.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "527976a0-c9c0-4eef-878d-b2aeee8427e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3) Keep only genes the (now-renamed) mouse matrix shares with the human one\n",
    "common = adata_muris.var_names.intersection(adata_sapiens.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5aa417b2-e3e1-49e1-b486-57a482fecc49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15383"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(common)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f6b87a54-656f-4721-a029-d0b0065cd867",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_muris  = adata_muris[:, common]\n",
    "adata_sapiens  = adata_sapiens[:, common]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5fcdb163-be7c-4fd3-b729-f58ac3dafe21",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_muris = adata_muris[:, adata_sapiens.var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "64c6abd5-434f-4f74-8d28-02e6f8fbfd68",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_names = ['cell_type', 'development_stage', 'sex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "087b234c-21ed-4974-b9e8-23509a1e5e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert adata_muris.var_names.tolist() == adata_sapiens.var_names.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "35f29c3b-6837-4e16-b5e3-d41e4513d365",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "classical monocyte                                      7922\n",
       "bronchial smooth muscle cell                            2339\n",
       "intermediate monocyte                                   1749\n",
       "fibroblast of lung                                      1523\n",
       "B cell                                                  1501\n",
       "alveolar macrophage                                     1366\n",
       "natural killer cell                                     1193\n",
       "lung macrophage                                         1157\n",
       "non-classical monocyte                                  1010\n",
       "CD8-positive, alpha-beta T cell                          870\n",
       "neutrophil                                               552\n",
       "CD4-positive, alpha-beta T cell                          551\n",
       "adventitial cell                                         526\n",
       "mature NK T cell                                         420\n",
       "vein endothelial cell                                    320\n",
       "T cell                                                   251\n",
       "myeloid dendritic cell                                   245\n",
       "pulmonary interstitial fibroblast                        222\n",
       "basophil                                                 130\n",
       "pulmonary alveolar type 2 cell                           125\n",
       "regulatory T cell                                        123\n",
       "smooth muscle cell of the pulmonary artery                98\n",
       "plasmacytoid dendritic cell                               75\n",
       "pericyte                                                  61\n",
       "multiciliated columnar cell of tracheobronchial tree      57\n",
       "plasma cell                                               49\n",
       "dendritic cell                                            45\n",
       "endothelial cell of lymphatic vessel                      41\n",
       "club cell                                                 15\n",
       "pulmonary neuroendocrine cell                              4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_muris.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "40d380ae-0594-4f22-9d0e-41ff2453dc65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "macrophage                                16486\n",
       "pulmonary alveolar type 2 cell            11594\n",
       "capillary endothelial cell                 7243\n",
       "basal cell                                 4015\n",
       "pulmonary alveolar type 1 cell             3116\n",
       "intermediate monocyte                      2785\n",
       "CD4-positive, alpha-beta T cell            2124\n",
       "CD8-positive, alpha-beta T cell            1898\n",
       "endothelial cell of artery                 1754\n",
       "club cell                                  1747\n",
       "classical monocyte                         1569\n",
       "vein endothelial cell                      1336\n",
       "basophil                                   1322\n",
       "lung multiciliated epithelial cell         1209\n",
       "alveolar adventitial fibroblast            1113\n",
       "respiratory tract goblet cell              1040\n",
       "natural killer cell                        1019\n",
       "pericyte                                    739\n",
       "B cell                                      663\n",
       "adventitial cell                            581\n",
       "non-classical monocyte                      527\n",
       "monocyte                                    506\n",
       "neutrophil                                  371\n",
       "endothelial cell of lymphatic vessel        315\n",
       "bronchial smooth muscle cell                220\n",
       "mature NK T cell                            163\n",
       "plasma cell                                 148\n",
       "vascular associated smooth muscle cell      131\n",
       "myeloid dendritic cell                       33\n",
       "pulmonary ionocyte                           25\n",
       "plasmacytoid dendritic cell                  19\n",
       "mesothelial cell                             18\n",
       "serous cell of epithelium of bronchus        15\n",
       "mast cell                                     3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sapiens.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d62be6ad-c651-4904-a563-ba1293f34c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mice_ctype = adata_muris.obs['cell_type'].unique()\n",
    "human_ctype = adata_sapiens.obs['cell_type'].unique()\n",
    "len(mice_ctype), len(human_ctype)\n",
    "\n",
    "common_ctypes = set(mice_ctype) & set(human_ctype)\n",
    "common_ctypes = list(common_ctypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "506ad90b-4159-4f04-b5d7-e004e66597fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(common_ctypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ae4159da-03ac-49e5-a206-9be02f8e9e50",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_muris = adata_muris[adata_muris.obs['cell_type'].isin(common_ctypes)]\n",
    "adata_sapiens = adata_sapiens[adata_sapiens.obs['cell_type'].isin(common_ctypes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2975bc71-bea9-4766-bee4-a1dbd03fd84f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 19694 × 15383\n",
       "    obs: 'age', 'cell', 'free_annotation', 'method', 'donor_id', 'n_genes', 'subtissue', 'n_counts', 'louvain', 'leiden', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'n_cells', 'means', 'dispersions', 'dispersions_norm', 'highly_variable', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type'\n",
       "    uns: 'age_colors', 'citation', 'leiden', 'louvain', 'neighbors', 'organism', 'organism_ontology_term_id', 'pca', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_muris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "70ab7aeb-1093-4a33-9cdf-9bfca43b4ae5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5659.0, 0.0)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_muris.X.max(), adata_muris.X.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6eda9a30-2569-4360-915d-2204efd87c73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2491817.0, 0.0)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sapiens.X.max(), adata_sapiens.X.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "07e72fb6-de79-4a52-a7b7-eea72efbb1f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_107091/789124555.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sapiens.obs['specie'] = 'sapiens'\n",
      "/tmp/ipykernel_107091/789124555.py:2: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_muris.obs['specie'] = 'muris'\n"
     ]
    }
   ],
   "source": [
    "adata_sapiens.obs['specie'] = 'sapiens'\n",
    "adata_muris.obs['specie'] = 'muris'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d6832979-e149-470f-91e4-68b9f4f3c107",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a6b9a342-e70f-4d99-b834-fd41ac901175",
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_key = 'cell_type'\n",
    "condition_key = 'specie'\n",
    "stim_name = 'sapiens'\n",
    "control_name = 'muris'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6d987cb4-0c2c-4016-880a-f5409fa71fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ad.concat([adata_sapiens, adata_muris])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b7144f10-9335-42e1-b569-419a2be3331d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/SE/miniconda3/envs/trials/lib/python3.10/site-packages/numba/np/ufunc/parallel.py:371: NumbaWarning: \u001b[1mThe TBB threading layer requires TBB version 2021 update 6 or later i.e., TBB_INTERFACE_VERSION >= 12060. Found TBB_INTERFACE_VERSION = 12050. The TBB threading layer is disabled.\u001b[0m\n",
      "  warnings.warn(problem)\n"
     ]
    }
   ],
   "source": [
    "adata.layers['counts'] = adata.X.copy()\n",
    "adata.raw = adata.copy()\n",
    "adata.uns = {}\n",
    "\n",
    "sc.pp.filter_cells(adata, min_counts=100)\n",
    "sc.pp.filter_genes(adata, min_counts=5)\n",
    "\n",
    "sc.pp.normalize_total(\n",
    "    adata, \n",
    "    target_sum=1e4, \n",
    "    exclude_highly_expressed=True\n",
    "    )\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "79f32f90-8b14-410b-8e3c-0331ac96e675",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 48867 × 15195\n",
       "    obs: 'donor_id', 'method', 'assay_ontology_term_id', 'cell_type_ontology_term_id', 'free_annotation', 'suspension_type', 'tissue_type', 'disease_ontology_term_id', 'is_primary_data', 'tissue_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'specie', 'n_counts'\n",
       "    var: 'n_counts'\n",
       "    uns: 'log1p'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ce7d6750-74b1-49b1-9a29-e10898d22121",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(\n",
    "    adata, \n",
    "    n_top_genes=5000, \n",
    "    subset=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "afface17-d1b7-408d-859a-9625e85ac83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['cov_cond'] = adata.obs[covariate_key].astype(str) + '_' + adata.obs[condition_key].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "01300b77-072e-4ecd-a317-f14a10f23032",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TSP2_Lung_proxmedialdistal_SS2_B113373_B134458_Immune_D2     non-classical monocyte_sapiens\n",
       "TSP2_Lung_proxmedialdistal_SS2_B113373_B134458_Stromal_I2                neutrophil_sapiens\n",
       "TSP2_Lung_proxmedialdistal_SS2_B113373_B134458_Empty_M12                 neutrophil_sapiens\n",
       "TSP2_Lung_proxmedialdistal_SS2_B113373_B134458_Stromal_P8                neutrophil_sapiens\n",
       "TSP2_Lung_proxmedialdistal_SS2_B113373_B134458_Stromal_P2               plasma cell_sapiens\n",
       "                                                                          ...              \n",
       "10X_P8_13_TTTATGCCATCCAACA-1                                      natural killer cell_muris\n",
       "10X_P8_13_TTTCCTCGTAAGGGAA-1                                         adventitial cell_muris\n",
       "10X_P8_13_TTTGCGCTCAAACCAC-1                                      natural killer cell_muris\n",
       "10X_P8_13_TTTGGTTTCCTGTACC-1                                      natural killer cell_muris\n",
       "10X_P8_13_TTTGTCACATATGAGA-1                                         adventitial cell_muris\n",
       "Name: cov_cond, Length: 48867, dtype: object"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cov_cond']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f18cb19f-d89b-458f-b10d-e9954e9e30dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "stim_degs = pt.compute_degs(\n",
    "    adata, \n",
    "    cov_key=covariate_key, \n",
    "    cond_key=condition_key, \n",
    "    stim_name=stim_name, \n",
    "    control_name=control_name,\n",
    "    condition_names=[control_name, stim_name],\n",
    "    method='wilcoxon'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "46eaa4a6-2bf9-463f-9371-278aafa6fccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[f'rank_genes_groups_{condition_key}'] = {\n",
    "    stim_name :stim_degs,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b3161dec-80be-41c4-8eac-addb97aced60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12.285428, 0.0)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.max(), adata.X.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "03fcbbb2-4734-4a1c-ac5c-f5896ac01d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['counts'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1c544f90-9097-4792-b53a-e4d839564a45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2491817.0, 0.0)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.max(), adata.X.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "63ae66c1-d151-4631-ae9e-4a6c8a1cd2d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_split = pt.create_split_cols(\n",
    "    adata=adata, \n",
    "    cov_key=covariate_key, \n",
    "    cond_key=condition_key, \n",
    "    stim_name=stim_name,\n",
    "    random_state=42\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "85c7d47e-e307-4192-b402-320e37f755c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_split.obs['sc_cell_ids'] = list(range(adata.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cbf986ec-70e1-4cb5-bdc7-5a4c12f12c72",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_split.write_h5ad('../../preprocessed_datasets/tabula.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "72ca07a4-6471-4862-874e-4e241e148f41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "pulmonary alveolar type 2 cell          11719\n",
       "classical monocyte                       9491\n",
       "intermediate monocyte                    4534\n",
       "CD8-positive, alpha-beta T cell          2768\n",
       "CD4-positive, alpha-beta T cell          2675\n",
       "bronchial smooth muscle cell             2559\n",
       "natural killer cell                      2212\n",
       "B cell                                   2164\n",
       "club cell                                1762\n",
       "vein endothelial cell                    1656\n",
       "non-classical monocyte                   1537\n",
       "basophil                                 1452\n",
       "adventitial cell                         1107\n",
       "neutrophil                                923\n",
       "pericyte                                  800\n",
       "mature NK T cell                          583\n",
       "endothelial cell of lymphatic vessel      356\n",
       "myeloid dendritic cell                    278\n",
       "plasma cell                               197\n",
       "plasmacytoid dendritic cell                94\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_split.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "36bb308c-b3b8-4b3c-a1a3-da7e2e50f4ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 48867 × 5000\n",
       "    obs: 'donor_id', 'method', 'assay_ontology_term_id', 'cell_type_ontology_term_id', 'free_annotation', 'suspension_type', 'tissue_type', 'disease_ontology_term_id', 'is_primary_data', 'tissue_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'specie', 'n_counts', 'cov_cond', 'split_sapiens_B cell', 'split_sapiens_CD4-positive, alpha-beta T cell', 'split_sapiens_CD8-positive, alpha-beta T cell', 'split_sapiens_adventitial cell', 'split_sapiens_basophil', 'split_sapiens_bronchial smooth muscle cell', 'split_sapiens_classical monocyte', 'split_sapiens_club cell', 'split_sapiens_endothelial cell of lymphatic vessel', 'split_sapiens_intermediate monocyte', 'split_sapiens_mature NK T cell', 'split_sapiens_myeloid dendritic cell', 'split_sapiens_natural killer cell', 'split_sapiens_neutrophil', 'split_sapiens_non-classical monocyte', 'split_sapiens_pericyte', 'split_sapiens_plasma cell', 'split_sapiens_plasmacytoid dendritic cell', 'split_sapiens_pulmonary alveolar type 2 cell', 'split_sapiens_vein endothelial cell', 'sc_cell_ids'\n",
       "    var: 'n_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'log1p', 'hvg', 'rank_genes_groups_specie'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_split"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (trials)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
